smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,205 @@
+import software.amazon.smithy.kotlin.codegen.lang.KotlinTypes
+import software.amazon.smithy.kotlin.codegen.model.boxed
+import software.amazon.smithy.kotlin.codegen.utils.doubleQuote
+import software.amazon.smithy.kotlin.codegen.utils.getOrNull
+import software.amazon.smithy.rulesengine.language.EndpointRuleset
+import software.amazon.smithy.rulesengine.language.eval.Value
+import software.amazon.smithy.rulesengine.language.lang.Identifier
+import software.amazon.smithy.rulesengine.language.lang.parameters.Parameter
+import software.amazon.smithy.rulesengine.language.lang.parameters.ParameterType
+
+private const val DEFAULT_DEPRECATED_MESSAGE =
+    "This field is deprecated and no longer recommended for use."
+
+private const val CLASS_NAME = "EndpointParameters"
+
+/**
+ * Renders the struct of parameters to be passed to the endpoint provider for resolution.
+ */
+class EndpointParametersGenerator(
+    private val writer: KotlinWriter,
+    rules: EndpointRuleset,
+) {
+    private val params: List<KotlinEndpointParameter> = rules.parameters.toList().map {
+        KotlinEndpointParameter(
+            it.name.toKotlin(),
+            it.type.toSymbol(),
+            it.isRequired,
+            it.defaultValue.getOrNull()?.toKotlinLiteral() ?: "null",
+            it.documentation.getOrNull(),
+            it.deprecated.getOrNull(),
+        )
+    }
+
+    fun render() {
+        renderDocumentation()
+        writer.withBlock("public class #L private constructor(builder: Builder) {", "}", CLASS_NAME) {
+            renderFields()
+            renderCompanionObject()
+            write("")
+            if (params.any(KotlinEndpointParameter::isRequired)) {
+                renderInit()
+                write("")
+            }
+            renderEquals()
+            write("")
+            renderHashCode()
+            write("")
+            renderToString()
+            write("")
+            renderCopy()
+            write("")
+            renderBuilder()
+        }
+    }
+
+    private fun renderFields() {
+        params.forEach {
+            writer.ensureSuppressDeprecation(it)
+            it.renderDeclaration(writer, "builder.${it.name}")
+            writer.write("")
+        }
+    }
+
+    private fun renderDocumentation() {
+        writer.dokka {
+            write("The set of values necessary for endpoint resolution.")
+        }
+    }
+
+    private fun renderCompanionObject() {
+        writer.withBlock("public companion object {", "}") {
+            write("public operator fun invoke(block: Builder.() -> Unit): #L = Builder().apply(block).build()", CLASS_NAME)
+        }
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""requireNotNull(#1L) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
```

Comment:
**Question**: In `StructureGenerator`, we do this at the field declaration site (e.g., `public val foo: String = requireNotNull(builder.foo) { "..." }`). Is there any reason to favor using an explicit `init` block for this?

Reply:
Agreed. Also if we do keep an `init{...}` block for whatever reason I'd consider moving it right after field declaration (before the companion object)